server/service/service.go

```diff
@@ -404,7 +405,64 @@ func (s *Service) BootstrapStream(stream bpb.Bootstrap_BootstrapStreamServer) er
 			}
 
 		case *bpb.BootstrapStreamRequest_Response_:
-			// TODO: process response
+			log.Infof("Received Response from %s", deviceID)
+			if session.currentState != stateTPM20NonceSent {
+				return status.Errorf(codes.FailedPrecondition, "unexpected state %v for device %s, expecting nonce response", session.currentState, deviceID)
+			}
+
+			challengeResponse := req.Response
+
+			nonceRespWrapper, ok := challengeResponse.Type.(*bpb.BootstrapStreamRequest_Response_NonceSigned)
+			if !ok {
+				return status.Errorf(codes.InvalidArgument, "expecting nonce challenge response type, got %T", challengeResponse.Type)
+			}
+			signedNonce := nonceRespWrapper.NonceSigned
+
+			// Verify the signed nonce using the public key from the device's IDevID certificate.
+			publicKey, ok := session.idevidCert.PublicKey.(*rsa.PublicKey)
+			if !ok {
+				return status.Errorf(codes.InvalidArgument, "public key from idevid_cert is not of type RSA")
+			}
+			hasher := crypto.SHA256.New()
+			hasher.Write(session.nonce)
+			hashedNonce := hasher.Sum(nil)
+
+			if err := rsa.VerifyPKCS1v15(publicKey, crypto.SHA256, hashedNonce, signedNonce); err != nil {
+				return status.Errorf(codes.PermissionDenied, "nonce signature verification failed: %v", err)
```

Comment:
Permission denied probably isn't the right error code here. Internal or invalid argument is better.

Reply:
done